drivers/clock_control/clock_control_mcux_ccm_rev3.c

```diff
@@ -0,0 +1,555 @@
+	if ((on && clk->state == IMX_CCM_CLOCK_STATE_UNGATED) ||
+	    (!on && clk->state == IMX_CCM_CLOCK_STATE_GATED)) {
+		return 0;
+	}
+
+	ret = imx_ccm_on_off(dev, clk, on);
+	if (ret < 0) {
+		LOG_ERR("failed to gate/ungate clock %s: %d", clk->name, ret);
+		return ret;
+	}
+
+	switch (clk->state) {
+	case IMX_CCM_CLOCK_STATE_GATED:
+		clk->state = IMX_CCM_CLOCK_STATE_UNGATED;
+		break;
+	case IMX_CCM_CLOCK_STATE_UNGATED:
+		clk->state = IMX_CCM_CLOCK_STATE_GATED;
+		break;
+	default:
+		/* this should never happen */
+		__ASSERT(false, "invalid clock state: %d", clk->state);
+	};
+
+	return 0;
+}
+
+static int _mcux_ccm_on(const struct device *dev,
+			struct imx_ccm_clock *clk)
+{
+	int ret;
+
+	LOG_DBG("currently ungating clock %s", clk->name);
+
+	if (!clk->parent) {
+		goto out_ungate;
+	}
+
+	ret = _mcux_ccm_on(dev, clk->parent);
+	if (ret < 0) {
+		LOG_ERR("failed ungating operation for clock %s", clk->parent->name);
+		return ret;
+	}
+
+out_ungate:
+	return mcux_ccm_on_off(dev, clk, true);
+}
+
+static int mcux_ccm_on(const struct device *dev, clock_control_subsys_t sys)
+{
+	struct imx_ccm_clock *clk;
+	int ret;
+
+	ret = imx_ccm_get_clock(dev, POINTER_TO_UINT(sys), &clk);
+	if (ret < 0) {
+		LOG_ERR("failed to get clock data for 0x%lx: %d",
+			POINTER_TO_UINT(sys), ret);
+		return ret;
+	}
+
+	/* ungate current clock and its parents */
+	return _mcux_ccm_on(dev, clk);
+}
+
+static int mcux_ccm_off(const struct device *dev, clock_control_subsys_t sys)
+{
+	struct imx_ccm_clock *clk;
+	int ret;
+
+	ret = imx_ccm_get_clock(dev, POINTER_TO_UINT(sys), &clk);
+	if (ret < 0) {
+		LOG_ERR("failed to get clock data for 0x%lx: %d",
+			POINTER_TO_UINT(sys), ret);
+		return ret;
+	}
+
+	return mcux_ccm_on_off(dev, clk, false);
+}
+
+static int mcux_ccm_get_rate(const struct device *dev,
+			     clock_control_subsys_t sys, uint32_t *rate)
+{
+	const struct imx_ccm_config *cfg;
+	struct imx_ccm_data *data;
+	struct imx_ccm_clock *clk;
+	int ret;
+
+	cfg = dev->config;
+	data = dev->data;
+
+	ret = imx_ccm_get_clock(dev, POINTER_TO_UINT(sys), &clk);
+	if (ret < 0) {
+		LOG_ERR("failed to get clock data for 0x%lx: %d",
+			POINTER_TO_UINT(sys), ret);
+		return ret;
+	}
+
+	/* clock not configured yet */
+	if (!clk->freq) {
+		LOG_ERR("can't get rate of unconfigured clock %s: %d",
+			clk->name, ret);
+		return -EINVAL;
+	}
+
+	*rate = clk->freq;
+
+	return 0;
+}
+
+static int _mcux_ccm_set_rate(const struct device *dev,
+			      struct imx_ccm_clock *clk,
+			      uint32_t rate)
+{
+	uint32_t parent_rate;
+	int ret, obtained_rate, clk_state;
+
+	clk_state = clk->state;
+
+	LOG_DBG("trying to set rate %u for clock %s", rate, clk->name);
+
+	/* note: although a set_clock_rate() operation may not
+	 * yield a frequency equal to the requested rate, this
+	 * will help filter out the cases in which it does.
+	 *
+	 */
+	if (clk->freq == rate) {
+		LOG_ERR("clock %s already set to rate %u", clk->name, rate);
+		return -EALREADY;
```

Comment:
Why return an error here? The clock is at the rate we need, so I would argue the function can just `return 0` here, since the clock is configured how the user needs it to be

Reply:
According to https://docs.zephyrproject.org/latest/hardware/peripherals/clock_control.html, set_rate() should return -EALREADY if clock was already configured to requested frequency. Despite this, I agree with you but we should try to abide by the clock controller API specification.